### PR TITLE
Refresh a subset of bundles if they provide missing requirements

### PR DESCRIPTION
**Pull request by laeubi**
When refreshing large sets of bundles the resolver sometimes take a choice where a requirement can't be bound even though it is there and resolvable.

This adds some additional code to detect this situation and then refresh a smaller set of bundles that provide these until a max retry is reached, all bundles are resolved, or all providers have been tried to refresh already.

### CODE REVIEW

File: bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/ConfigApplier.java
Comment (tjwatson) [inline bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/ConfigApplier.java:132]: Can you explain why this would need to be tried more than once?  If it didn't resolve after the first attempt to refresh providers, why would it work on subsequent retries?

File: bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/ConfigApplier.java
Comment (tjwatson) [inline bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/ConfigApplier.java:124]: Do you have a scenario that exhibits this behavior?  Is there going to be a way to add a test for this?

File: bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/Activator.java
Comment (tjwatson) [inline bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/Activator.java:46]: I'm unsure why this doesn't use the `org.eclipse.osgi.service.debug.DebugOptions` service to get trace options for the simple configurator.  Probably because of the ill-fated dream to have p2 be able to run on other framework implementations.

**Review by tjwatson — COMMENTED** [review]
The code is fine given what it is trying to do.  But I question if we can have a testcase that exhibits the behavior so we can add a test for this code path.

**Comment by laeubi** [comment]
> The code is fine given what it is trying to do. But I question if we can have a testcase that exhibits the behavior so we can add a test for this code path.

I'm not yet sure in wich form such a test-case would have to be provided, can I somehow "export" a bundle-set (metadata-only) and use it as an input for a test with simple-configurator?

File: bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/Activator.java
Comment (laeubi) [reply bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/Activator.java:46]: I could think about that tracing might not be available at this stage? But not completely sure, that way you can at least enable it quite easy when starting Eclipse from commandline.

File: bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/ConfigApplier.java
Comment (laeubi) [reply bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/ConfigApplier.java:124]: Yes I have a product where I install a larger set of plugins into an older Eclipse version, but I'm not sure how this cane be transformed into a standalone test-case, do we have such facility somewhere?

I could think about exporting all bundles into a OSGi Repository and then call the resolver api with this repository as a source...

File: bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/Activator.java
Comment (tjwatson) [reply bundles/org.eclipse.equinox.simpleconfigurator/src/org/eclipse/equinox/internal/simpleconfigurator/Activator.java:46]: We can leave that to another issue

**Comment by tjwatson** [comment]
@laeubi you may not like this, but to move this forward I would like to have an option that defaults to the current behavior so we can enable this for the problematic scenarios.  I would rather do that now so that the ones impacted by this issue can try out the solution without risking others.  Once we get more confident in the approach then we can flip the default or get rid of the option.

**Comment by laeubi** [comment]
@tjwatson do you like to suggest how this should be implemented? Maybe as simple as a system property to check?

**Comment by tjwatson** [comment]
> @tjwatson do you like to suggest how this should be implemented? Maybe as simple as a system property to check?

Yes, but use `BundleContext.getProperty` instead of `System.getProperty`.

**Comment by laeubi** [comment]
@tjwatson I now added this inside a block that is disabled by default and can be enabled with `equinox.simpleconfigurator.deeprefresh=true`

**Review by tjwatson — APPROVED** [review]
LGTM,

Thanks for putting the option to enable here.  As developers we should encourage committers and contributors to test with this enabled so we can evaluate enabling this by default in the future.

**Pull request merged**
